applications/FluidDynamicsApplication/custom_elements/data_containers/derivatives.h

```diff
@@ -0,0 +1,172 @@
+
+    static constexpr unsigned int BlockSize = SubAssemblyType::BlockSize;
+
+    static constexpr unsigned int ResidualSize = NumNodes * BlockSize;
+
+    ///@}
+    ///@name Public operations
+    ///@{
+
+    inline BoundedVector<double, ResidualSize>& GetSubVector()
+    {
+        return mSubVector;
+    }
+
+    template<class TCombinedElementDataContainer>
+    inline typename std::tuple_element<ElementDataIndex, TCombinedElementDataContainer>::type& GetElementDataContainer(TCombinedElementDataContainer& rCombinedDataContainer) const
+    {
+        static_assert(
+            ElementDataIndex < std::tuple_size<TCombinedElementDataContainer>::value,
+            "Required Element data container index is more than the available element data containers.");
+        return std::get<ElementDataIndex>(rCombinedDataContainer);
+    }
+
+    template<IndexType TAssemblyRowBlockSize, IndexType TAssemblyColumnBlockSize = TAssemblyRowBlockSize>
+    inline void AssembleSubVectorToMatrix(Matrix& rOutput, const IndexType NodeIndex) const
+    {
+        KRATOS_TRY
+
+        KRATOS_DEBUG_ERROR_IF(rOutput.size1() != TAssemblyRowBlockSize * NumNodes)
+            << "rOuput.size1() does not have the required size. [ rOutput.size1() = "
+            << rOutput.size1() << ", required_size = TAssemblyRowBlockSize * NumNodes = "
+            << TAssemblyRowBlockSize * NumNodes << " ].\n";
+
+        KRATOS_DEBUG_ERROR_IF(rOutput.size2() != TAssemblyColumnBlockSize * NumNodes)
+            << "rOuput.size2() does not have the required size. [ rOutput.size2() = "
+            << rOutput.size2() << ", required_size = TAssemblyColumnBlockSize * NumNodes = "
+            << TAssemblyColumnBlockSize * NumNodes << " ].\n";
+
+        for (IndexType i = 0; i < NumNodes; ++i) {
+            for (IndexType j = 0; j < BlockSize; ++j) {
+                rOutput(TAssemblyRowBlockSize * NodeIndex + RowStartingIndex, i * TAssemblyColumnBlockSize + j + ColumnStartingIndex) += mSubVector[i * BlockSize + j];
+            }
+        }
+
+        KRATOS_CATCH("");
+    }
+
+    template<IndexType TAssemblyRowBlockSize>
+    inline void AssembleSubVectorToVector(Vector& rOutput) const
+    {
+        KRATOS_TRY
+
+        static_assert(ColumnStartingIndex == 0);
+
+        KRATOS_DEBUG_ERROR_IF(rOutput.size() != TAssemblyRowBlockSize * NumNodes)
+            << "rOuput.size() does not have the required size. [ rOutput.size() = "
+            << rOutput.size() << ", required_size = TAssemblyRowBlockSize * NumNodes = "
+            << TAssemblyRowBlockSize * NumNodes << " ].\n";
+
+        for (IndexType i = 0; i < NumNodes; ++i) {
+            for (IndexType j = 0; j < BlockSize; ++j) {
+                rOutput[i * TAssemblyRowBlockSize + j] += mSubVector[i * BlockSize + j];
+            }
+        }
+
+        KRATOS_CATCH("");
+    }
+
+    ///@}
+private:
+    ///@name Private members
+    ///@{
+
+    BoundedVector<double, ResidualSize> mSubVector;
```

Comment:
(please correct me if I'm wrong, I might have missunderstood the purpose of this class)

This `mSubVector` is a copy of the vector you want to assemble. If so it is, wouldn't it be better to make the two assembling functions `AssembleSubVectorToVector` and `AssembleSubVectorToMatrix` `static` and passing the vector as an argument?. Aside of the copy, this will avoid calling the `GetSubVector` to assign the value of `mSubVector` before calling the assembly methods.

Reply:
OK. I wasn't aware of these implications.